lib/auth/auth_with_roles.go

```diff
@@ -1461,8 +1461,25 @@ func (a *ServerWithRoles) NewWatcher(ctx context.Context, watch types.Watch) (ty
 				return nil, trace.Wrap(err)
 			}
 
-			// Only users can watch their own headless authentications.
-			if !hasLocalUserRole(a.context) || filter.Username != a.context.User.GetName() {
+			// Users can only watch their own headless authentications.
+			if !hasLocalUserRole(a.context) {
+				if watch.AllowPartialSuccess {
+					continue
+				}
+				return nil, trace.AccessDenied("non-local user roles cannot watch headless authentications")
+			}
+
+			if filter.Username == "" {
+				if watch.AllowPartialSuccess {
+					continue
+				}
+				return nil, trace.AccessDenied("user cannot watch headless authentications without a filter for their username")
+			}
+
+			if filter.Username != a.context.User.GetName() {
+				if watch.AllowPartialSuccess {
```

Comment:
Since the scope of the watcher here is used for authentication, I am curious to understand what scenario may trigger (here and above) the need for `AllowPartialSuccess` ?

Reply:
I can't speak to the exact use case of `AllowPartialSuccess`, but generally it allows a watcher with multiple watch kinds specified to succeed, even if it can only watch a subset of the watch kinds specified. Unauthorized watch kinds do not get watched in this case as they are not added to the list of `validKinds` below, so there is no risk of authz bypass.

https://github.com/gravitational/teleport/blob/a3d89ff6ce231139a3a58f7becb09a122a12b67d/lib/auth/auth_with_roles.go#L1494

Reply:
Noted about the `validKinds`. Even then, I would be inclined to remove `AllowPartialSuccess` altogether if we do not need such partial matches for headless. Your call though!